taichi/ir/control_flow_graph.cpp

```diff
@@ -203,19 +227,31 @@ Stmt *CFGNode::get_store_forwarding_data(Stmt *var, int position) const {
     }
     return true;  // continue the following loops
   };
+
+  // [Global Addr only]
+  // if there's a store to the same dest_addr in a previous block
+  // if the store values are the same, then return the value
+  last_def_position = -1;
   for (auto stmt : reach_in) {
     // var == stmt is for the case that a global ptr is never stored.
     // In this case, stmt is from nodes[start_node]->reach_gen.
     if (var == stmt || may_contain_address(stmt, var)) {
       if (!update_result(stmt))
         return nullptr;
+      else
+        last_def_position = 0;
```

Comment:
Is the value here simply an integer that is neither positive nor -1?

Reply:
It is not, the background is that whenever we found a `load` - `store` pair that loads/stores to the same address, for example:
```
$1: store $0, val
....
$2 = load $0
```

We will then loop through and analyze the statements in between (those in the `....`) to check if it's same to perform the load-to-store forwarding. The validation logic is included in `may_contain_address()`.

Now, if the `load` - `store` pair is not in the same node (roughly one node corresponds to one code block), then the `$1: store $0, val` is likely gonna stay in another node, and its value tracked in `reach_in`. In this case, the statements we need to validate should start from the very beginning of this node all the way down to where `$2 = load $0` locates. That is why I put `last_def_position = 0` here.

Reply:
Thank you for the clarification! I now have a clear understanding that 0 represents the minimal valid index within a node.